### PR TITLE
Steve/remove rust add submodule

### PR DESCRIPTION
**Pull request by qwang98**
Also broken up the files and modified some features to avoid circular references.

### CODE REVIEW

**Comment by leolara** [comment]
@qwang98 I don't think we need the Cargo stuff in the root anymore, as the submodule contains it

**Comment by leolara** [comment]
@qwang98 The code in lib.rs should also be in rust project frontend/pychiquito.rs

**Comment by qwang98** [comment]
@qwang98 I don't think we need the Cargo stuff in the root anymore, as the submodule contains it

The line in Cargo.toml chiquito = { path = "./src/chiquito" } is actually using the submodule as it points to a local directory. I tried just doing pub mod chiquito or pub mod src/chiquito in lib.rs but won't work, because the chiquito files are contained in the src folder. Let me know if you have a better way though.

**Comment by qwang98** [comment]
> @qwang98 The code in lib.rs should also be in rust project frontend/pychiquito.rs

Currently we are using `maturin` to build the Rust bindings to Python module and automatically link it to the Python virtual environment project. Its setup requires exposing the bindings in `lib.rs`, so we can't really move them to `pychiquito.rs` if we use `maturin`. I tried moving them but `maturin` couldn't build the Rust bindings any more. See the setup of `maturin` here: https://pyo3.rs/v0.19.1/#using-rust-from-python

Alternatively, we can also build the Rust project manually. This would require building the Rust bindings in the Rust submodule, then renaming and moving it from Rust /target directory to the output directory for Python shell to import. If we were to do this, we should make some scripts so that the user don't need to do this. See guide: https://pyo3.rs/v0.19.1/building_and_distribution.html#manual-builds.

Let me know which one of the two options we want. In summary:
1. Keep using `maturin` and keep the Rust bindings in `lib.rs` rather than Rust chiquito.
2. Use manual build and write a script to build and move the bindings to an output directory for Python.

**Comment by leolara** [comment]
@qwang98 please use this for the UUID generation https://docs.python.org/3/library/uuid.html

Check how it is in the rust project:

https://github.com/privacy-scaling-explorations/chiquito/blob/main/src/util.rs

**Comment by leolara** [comment]
@qwang98 the python code cannot be in .env, it shouild be in a directory like pychiquito for example

File: .env/src/cb.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        annotation: str = str(expr)
+        if isinstance(expr, StepTypeNext):
+            return Constraint(annotation, expr, Typing.Boolean)
+        else:
+            return Constraint(annotation, expr, Typing.Unknown)
+
+    def __str__(self: Constraint) -> str:
+        return self.annotation
+
+    def cb_and(
```
Comment (leolara) [inline .env/src/cb.py]: @qwang98 will this mean we need to call Constraint.cb_and(....) ? Can we overload the and operator? Or is there a way to make it less verbose?

Comment (leolara) [reply .env/src/cb.py]: Is it necessary that this functions have to be members of the Constraint class? Cannot be simple functions?

Comment (qwang98) [reply .env/src/cb.py]: I don't think it's possible to overload `and` `or` `not` in Python. You can overload `&` `|` with `__and__` `__or__`, but can't overload the literal `and` `or`.

They can be simple functions and I will change all of them to behave like that!

**Comment by qwang98** [comment]
> @qwang98 the python code cannot be in .env, it shouild be in a directory like pychiquito for example

This is something specific to `maturin`, which uses virtual environment to auto setup Rust binding generation and apply them to Python. See my comment above on `maturin`.

However, I also suggested potentially manually setting the bindings up with some scripts, though I cannot guarantee that Python code can move out of `.env`. Let me know which way we want to proceed with!

**Comment by leolara** [comment]
Regarding maturin, ok, let's have it now like this. But, in the future all functions in src/lib.rs should be very simple and just call functions in frontend/pychiquito.rs

**Comment by qwang98** [comment]
> @qwang98 please use this for the UUID generation https://docs.python.org/3/library/uuid.html
> 
> Check how it is in the rust project:
> 
> https://github.com/privacy-scaling-explorations/chiquito/blob/main/src/util.rs

Fixed!

File: .env/src/cb.py (hunk 1 above)
Comment (qwang98) [reply .env/src/cb.py]: Fixed! Made all cb functions simple functions so we don't need to Constraint.eq, for example, any more.

**Comment by qwang98** [comment]
> Regarding maturin, ok, let's have it now like this. But, in the future all functions in src/lib.rs should be very simple and just call functions in frontend/pychiquito.rs

Got it. Then I'll make the frontend.rs call the boilerplate and then src/lib.rs will do a simple invoke of those frontend.rs functions.

**Comment by leolara** [comment]
> > Regarding maturin, ok, let's have it now like this. But, in the future all functions in src/lib.rs should be very simple and just call functions in frontend/pychiquito.rs
> 
> Got it. Then I'll make the frontend.rs call the boilerplate and then src/lib.rs will do a simple invoke of those frontend.rs functions.

frontend/pychiquito.rs, no frontend.rs

**Comment by qwang98** [comment]
> > > Regarding maturin, ok, let's have it now like this. But, in the future all functions in src/lib.rs should be very simple and just call functions in frontend/pychiquito.rs
> > 
> > 
> > Got it. Then I'll make the frontend.rs call the boilerplate and then src/lib.rs will do a simple invoke of those frontend.rs functions.
> 
> frontend/pychiquito.rs, no frontend.rs

Yes my typo lol! That's what I did in the new commit. :)

**Review by leolara — APPROVED** [review]

**Pull request merged**